slack_incoming: Handle optional attachment fields aptly.

prah23 opened the pull request:
This PR checks for null values for keys within "attachment" in
the Slack integration's incoming payloads. These keys were expected
to exist optionally previously, and the existence of null values for
these wasn't anticipated. Due to an issue report for such null
values in the payload, their handling is updated appropriately.

The checks for these values are truthiness checks since the strategy
for these values being null or falsy ("", 0) is the same; we don't
process that key-value pair. This is consistent with how Slack handles
this scenario.

For the case where all the attachment fields have null values, Slack
displays this as an empty block with no content, and therefore our
strategy for this is a no-op.

Tests updated.

<!-- Describe your pull request here.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

**Self-review checklist**

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/version-control.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.

asah commented:
@prah23 let me know if you need help with testing / unittests, I'm happy to
help.


On Wed, Aug 17, 2022 at 3:07 PM Tim Abbott ***@***.***> wrote:

> @prah23 <https://github.com/prah23> how did you test this? Ideally, we'd
> add a test case with a fixture captured from Slack that actually has one of
> these values None.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/zulip/zulip/pull/22738#issuecomment-1218389155>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABCFYXMH6DAT5KWJK7DYVTVZUZ5RANCNFSM562QCVQQ>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

prah23 commented:
@timabbott Thanks for the quick review!

I've used the same documentation that's already commented in the webhook's code. For `attachment` within the payload, the Slack documentation mentions that legacy fields are optional if the new `block` structure is being used. Apart from them being optional, the report we've received on CZO suggests those keys may exist and have `None` as a value as well. Therefore, I've added a check for each of these legacy fields.

Apart from this, I also went through the documentation for `block` and checked for any keys that have "Required?" mentioned as "No" (signifying them being optional) and added a `None` check for those keys as well.

I haven't tested this with a Slack webhook fixture yet, I opened this PR to track progress for this issue. I'll add a '[WIP]' and update it once I have updated tests with valid fixtures.

@asah Sure, thanks!

timabbott commented:
Merged, thanks @prah23!
